### PR TITLE
Pool of devices and async execution of multiple active devices supports

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
- The runtime now supports multiple active devices managed via a device pool. The new `RTDevice` data-class and `RTDeviceStatus` along with the `thread_local` device instance pointer enable the runtime to better scope the lifetime of device instances concurrently. With these changes, one can create multiple active devices and execute multiple programs in a multithreaded environment.

- Add `DeviceReleaseOp` to the Quantum MLIR dialect. This will be lowered to the `__quantum__rt__device_release()` runtime instruction updating the status of the device instance from `Active` to `Inactive`. The runtime will reuse this deactivated instance instead of creating a new one automatically at runtime in a  multi-qnode workflow when another device with identical specifications is requested.
 

[sc-47651]

### CODE REVIEW

File: doc/changelog.md
Comment (josh146) [inline doc/changelog.md:24]: Hey @maliasadi! What does this change mean for the end-user? Are there any features etc. it unlocks? If so, any code examples we could add here?

File: runtime/lib/capi/ExecutionContext.hpp
Comment (rmoyard) [inline runtime/lib/capi/ExecutionContext.hpp:358]: It is more create or get, right?

**Review by rmoyard — COMMENTED** [review]
It looks great! :100:  I have some questions

File: demos/adaptive_circuits_demo.ipynb
Comment (josh146) [inline demos/adaptive_circuits_demo.ipynb:1]: @maliasadi was this a bug in the demo?

File: runtime/lib/backend/openqasm/OpenQasmDevice.cpp
Comment (maliasadi) [reply runtime/lib/backend/openqasm/OpenQasmDevice.cpp:45]: The state of a device needs to be reset when the program releases all qubits. In terms of OQ3 devices, this is equivalent to assign a new instance of `OpenQasm::Builder` to the builder variable of the device.

File: mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp
Comment (erick-xanadu) [inline mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:104]: Is this re-added somewhere else? I am not yet intimately familiar with gradients so maybe @dime10 can chime in here. The concrete question is, we are not "leaking devices" by having this removing DeviceReleaseOp here, or are we?

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:113]: yes, the `lock_guard` is a RAII class around `mu.lock` and `mu.unlock`.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:206]: I added type at the end of this class to make it homogeneous with `RTDeviceStatus`. What would you prefer?
- `RTDeviceType` and `RTDeviceStatus`
- `RTDevice` and `RTDeviceStatus`

I don't have any strong preference here.

File: mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp
Comment (erick-xanadu) [reply mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp:104]: Nope, this is good. Never mind!

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:273]: Sure, `rtd_lib` is the absolute path to the device shared library (`/absolute/path/to/libdevice.so`) in string. We use `rtd_lib` to create an instance of `SharedLibraryManager`. This shared library manager is responsible to validate the library path, open and close the shared library, and get us the symbol to the factory method of the device. The `RTDeviceType` stores the factory method as well as `rtd_lib`. But, we don't really need to keep the share library manager around after getting the device factory method, so this will be destructed at the end of the scope of this method.

File: doc/changelog.md
Comment (dime10) [inline doc/changelog.md:45]:
```suggestion
* The runtime now supports multiple active devices managed via a device pool.
```

File: runtime/lib/capi/ExecutionContext.hpp
Comment (dime10) [inline runtime/lib/capi/ExecutionContext.hpp:368]: An alternative to this slightly confusing structure of pairs:

- `pool_counter` seems only used in order to hold device instances in a map. If we were to use a list instead (might need additional changes), that would remove the need for the first pair.
- Instead of keeping the status of a device as a pair in the pool, it also directly be stored in the `RTDevice` class. This would be a bit more OO and would eliminate the second pair.

I'm not saying we necessarily have to implement both of these, but they are some suggestions that could simplify the code.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:317]: This guard is used to avoid double initialization of the Python interpreter for Braket devices. If you call runtime from the frontend, we don't need to re-initialize the interpreter (using CPython functions). Otherwise,  we initialize a new Python interpreter instance in the runtime and keep it until the end of the program scope. This guard takes care of the lifetime of this interpreter instance.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:358]: This method returns a device pointer with its unique index via either creating a new device or reuse a released device. As it doesn't always create a new device instance, I didn't call it `createDevice` and instead it's named `getDevice`. Let me know what you think.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:400]: hehe... These C++ first/second field accesses make me crazy sometime. yep, I can simplify the code.. :+1:

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (dime10) [inline runtime/lib/capi/RuntimeCAPI.cpp:67]: I'm curious, if the execution context is global, why does it need to be passed in? Or are there potentially multiple instances of the execution context?

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (dime10) [inline runtime/lib/capi/RuntimeCAPI.cpp:48]: Would it make sense to have the thread_local variable point to the active `RTDevice` instance instead? That way we could eliminate the ID variable (I think).

**Review by dime10 — COMMENTED** [review]
Great work @maliasadi  🎉 🎉 

I mostly have some comments and questions to pin down the role of different components, but happy to approve otherwise :)

File: runtime/lib/capi/ExecutionContext.hpp
Comment (dime10) [reply runtime/lib/capi/ExecutionContext.hpp:400]: I have some suggestions for how this could be done here: https://github.com/PennyLaneAI/catalyst/pull/381#discussion_r1425825188

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (erick-xanadu) [inline runtime/lib/capi/RuntimeCAPI.cpp:42]: I think this should not be `thread_local`.

**Comment by erick-xanadu** [comment]
There's a frontend test in test_compiler that tests the writing a program using MLIR as a string. I think we need to add the new quantum instruction release there as well.

**Comment by erick-xanadu** [comment]
We also need to reconsider what happens now if the runtime throws an exception.

File: doc/changelog.md
Comment (maliasadi) [reply doc/changelog.md]: I meant qubits :)

**Review by erick-xanadu — APPROVED** [review]
@maliasadi, this is looking good! I will mark this as approved but I now think we should wait to merge it until the async support is more worked out. I would like the async support for finite difference to be working well. You mentioned you wanted this merge as it would facilitate my work but I think it is ok, I can always rebase :). I'll let you and @rmoyard coordinate on this. 

PS: don't forget to fix the frontend test in `test_compiler.py`

**Comment by maliasadi** [comment]
@erick-xanadu Thanks! There is no rush to merge this PR if it doesn't contribute to your progress. I'll apply the remaining suggestions today and will ping you for the rebase. Let's ensure end-to-end support before merging :100:

File: doc/changelog.md
Comment (dime10) [reply doc/changelog.md]: Why does device release release qubits? The whole point of this instruction was to separate that concern no?

File: mlir/include/Quantum/IR/QuantumOps.td
Comment (dime10) [reply mlir/include/Quantum/IR/QuantumOps.td]: I think that's better 👍

File: mlir/lib/Gradient/Transforms/GradMethods/ClassicalJacobian.cpp
Comment (dime10) [reply mlir/lib/Gradient/Transforms/GradMethods/ClassicalJacobian.cpp:288]: Yes I think conceptually it belongs here, but it certainly could be that no tests depend on it.

File: demos/adaptive_circuits_demo.ipynb
Comment (rmoyard) [reply demos/adaptive_circuits_demo.ipynb:280]: Thanks :+1:

File: runtime/lib/capi/ExecutionContext.hpp
Comment (rmoyard) [reply runtime/lib/capi/ExecutionContext.hpp:316]: Thanks!

File: runtime/lib/capi/ExecutionContext.hpp
Comment (rmoyard) [reply runtime/lib/capi/ExecutionContext.hpp:357]: Minor: maybe ``getOrCreate`` would be nice

**Review by rmoyard — COMMENTED** [review]
Great work @maliasadi! Thanks for answering my questions! I let you answer David's comments and questions (that I also share) and you can consider that is all good on my side :+1:

File: runtime/lib/capi/ExecutionContext.hpp
Comment (erick-xanadu) [reply runtime/lib/capi/ExecutionContext.hpp:399]: Yes, it is not ideal. I don't know if it is copied properly: I found that this is potentially a source of the issue with finite differences:

```
(gdb) down
#4  0x00007fff921e2a5d in Catalyst::Runtime::ExecutionContext::releaseDevice (this=0x55555aac3460, device_key=1) at /home/erick.ochoalopez/Code/cataliist/runtime/lib/capi/ExecutionContext.hpp:404
404	            it->second.second->releaseDevice();
(gdb) it->second
Undefined command: "it->second".  Try "help".
(gdb) p it->second
$4 = {first = Catalyst::Runtime::RTDeviceStatus::Active, second = std::shared_ptr<Catalyst::Runtime::RTDeviceType> (empty) = {get() = 0x0}}
```

Can RTDeviceType ever be empty?

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:273]: By manager, I mean `SharedLibraryManager` and this is destructed (after fetching the factory method) at the end of this method. I cannot see any opposition here :)

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:319]: Ah, that was part of my first implementation. It's removed from `RTDevice`. 

This `initial_tape_recorder_status` is used to keep the initial state of the tape recorder before initializing the device. In the adjoint-jacobian support, the compiler lowers the code in a way that it calls `__quantum__rt__toggle_recorder(true)` before `__quantum__rt__device_init`.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:357]: Fixed!

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:367]: The current pool design is outdated as we no longer store devices with their name as keys. That was my original plan for this support. We no longer need `pool_counter` nor `unordered_map`. I like both of your suggestions... updated the support!

File: doc/changelog.md
Comment (maliasadi) [reply doc/changelog.md]: Removed it from the support and changelog. I added the release qubits to the release-device instruction to ensure all qubits are releases before deactivating a device. This is no longer required because of our IR model that the release-device instruction always comes after releasing all qubits of the device.

File: mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp
Comment (maliasadi) [reply mlir/lib/Gradient/Transforms/GradMethods/PS_QuantumGradient.cpp:309]: Weird.. The clang-formater formats this and it doesn't let me to update it in anyways!

File: runtime/lib/capi/ExecutionContext.hpp
Comment (dime10) [reply runtime/lib/capi/ExecutionContext.hpp:271]: So the problem I see is with the statement that "the `SharedLibraryManager` is responsible to open and _close_ the shared library". If the `SharedLibraryManager` is destroyed right after fetching the factory method, how can it be responsible to close the shared library (whose lifetime needs to be much longer than that of the manager)?

A different, related question would be to ask, how and when is the shared library closed right now?

File: runtime/lib/capi/ExecutionContext.hpp
Comment (dime10) [inline runtime/lib/capi/ExecutionContext.hpp:386]:
```suggestion
    void deactivateDevice(RTDevice *RTD_PTR)
```

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (dime10) [inline runtime/lib/capi/RuntimeCAPI.cpp:163]: I guess there is nothing wrong with resetting the device pointer, but since it seems redundant to me I'm just curious if there was specific reasoning behind it :)

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (dime10) [inline runtime/lib/capi/RuntimeCAPI.cpp:53]: I asked this in previous comment as well, but I don't quite understand why the EC would be passed in as a variable if it is global (a singleton essentially)?

Comment (dime10) [reply runtime/lib/capi/RuntimeCAPI.cpp:53]: The return type (bool) and the function name (indicating a device might be returned) don't really match either.

Comment (dime10) [reply runtime/lib/capi/RuntimeCAPI.cpp:53]: I'm also curious about the the `friend` attribute in the EC class, what purpose does it serve?

**Review by dime10 — COMMENTED** [review]
Thanks @maliasadi, this looks great! A few last questions about the interaction of the CAPI and the execution context, then I'm happy with it 💯

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (dime10) [reply runtime/lib/capi/RuntimeCAPI.cpp]: I re-asked the question in my recent review, so I'm closing this one again.

In general, even if a comment seems resolved by recent changes, it's good to state why and how you think it is now resolved before closing.

File: Makefile
Comment (maliasadi) [reply Makefile]: Thanks! Fixed [316da81](https://github.com/PennyLaneAI/catalyst/pull/381/commits/316da81e5352f4d22e44a071eab2b45b45df2a47)

Comment (maliasadi) [reply Makefile]: hmm.. didn't want to change the existing commands (I don't know why?!! :D) Fixed [316da81](https://github.com/PennyLaneAI/catalyst/pull/381/commits/316da81e5352f4d22e44a071eab2b45b45df2a47)

File: runtime/tests/Test_LightningCoreQIS.cpp
Comment (maliasadi) [reply runtime/tests/Test_LightningCoreQIS.cpp]: yes, thanks! fixed it [316da81](https://github.com/PennyLaneAI/catalyst/pull/381/commits/316da81e5352f4d22e44a071eab2b45b45df2a47)

**Review by dime10 — APPROVED** [review]
Nice work! 🥇

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp]: We don't close the loaded library when we build the backend devices with OPENMP on Linux due to some conflicts between the OPENMP threads and dlclose. But on macOS, I ran into issues with destructing the SLM instance at the end of the method. Thus, I changed the scope to be destructed by the `RTDevice` destructor to get the support on macOS.

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (maliasadi) [reply runtime/lib/capi/RuntimeCAPI.cpp]: Oops, I thought I got rid of all these friends methods after removing the `thread_local` device ID.
- Passing the EC as a variable was part of the `friend` method requirement (removed in the updated version)
- Returning a boolean is to ensue that we successfully initialize `RTD_PTR`. 
- I was using `friend` methods to access the device pool and manipulate devices using private methods and variables of the EC class. (removed in the updated version)

Thanks for your questions. They helped me properly tidy up the support.

**Pull request merged**
